File: training-data/swift/matrix_chain_multiplication.swift
func matrixChainOrder(_ dims: [Int]) -> Int {
    let n = dims.count - 1
    guard n > 1 else { return 0 }

    var dp = Array(repeating: Array(repeating: 0, count: n), count: n)
    for length in 2...n {
        for i in 0...(n - length) {
            let j = i + length - 1
            dp[i][j] = Int.max
            for k in i..<j {
                let cost = dp[i][k] + dp[k + 1][j] + dims[i] * dims[k + 1] * dims[j + 1]
                if cost < dp[i][j] {
                    dp[i][j] = cost
                }
            }
        }
    }
    return dp[0][n - 1]
}

// Matrices: A1(40x20), A2(20x30), A3(30x10), A4(10x30)
let dims = [40, 20, 30, 10, 30]
print(matrixChainOrder(dims))

File: training-data/c/matrix_chain_multiplication.c
#include <limits.h>
#include <stdio.h>

/* dims has n+1 entries: matrix i has dimensions dims[i-1] x dims[i]. */
int matrix_chain_order(int dims[], int n) {
    int dp[n][n];

    for (int i = 1; i < n; i++) dp[i][i] = 0;

    for (int len = 2; len < n; len++) {
        for (int i = 1; i <= n - len; i++) {
            int j = i + len - 1;
            dp[i][j] = INT_MAX;
            for (int k = i; k < j; k++) {
                int cost = dp[i][k] + dp[k + 1][j] + dims[i - 1] * dims[k] * dims[j];
                if (cost < dp[i][j]) dp[i][j] = cost;
            }
        }
    }
    return dp[1][n - 1];
}

int main(void) {
    /* Matrices: 40x20, 20x30, 30x10, 10x30 -> dims = {40,20,30,10,30} */
    int dims[] = {40, 20, 30, 10, 30};
    int n = sizeof(dims) / sizeof(dims[0]);

    printf("minimum multiplication cost: %d\n", matrix_chain_order(dims, n));
    return 0;
}
